File: state_ide/impl/elements/HsmTransition.py
# hsm-ide project
# Copyright (C) 2022 Igor Krechetov
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import math
from PySide6.QtWidgets import QGraphicsItem
from PySide6.QtCore import QPointF, QLineF, Qt, QRectF
from PySide6.QtGui import QPen, QPolygonF, QPainterPath

from impl.elements.private.HsmElement import HsmElement
from impl.elements.private.ElementBoundaryGripItem import ElementGripItem

# TODO: recalculate start and end positions when related items were moved
class HsmTransition(HsmElement):
    # Consts
    SELECTION_DISTANCE = 2

    def __init__(self):
        self.fromElement = None
        self.toElement = None
        self.selectionPath = QPainterPath()
        self.linePath = QPolygonF()
        # NOTE: there are no grips at the beginning and and of the poly-line
        self.lineGrips = [None, None]
        self.debugShowSelectionPolygon = False

        super().__init__()
        self.setFlag(QGraphicsItem.ItemIsMovable, False)
        self.setFlag(QGraphicsItem.ItemIsSelectable, True)
        self.setZValue(0)
        self.updateBoundingRect()

    # -----------------------------------
    # bounding rect
    def updateBoundingRect(self):
        self.prepareGeometryChange()
        self.updateSelectionPolygon()
        self.outerRect = self.linePath.boundingRect().normalized().adjusted(-self.SELECTION_DISTANCE,
                                                                            -self.SELECTION_DISTANCE,
                                                                            self.SELECTION_DISTANCE,
                                                                            self.SELECTION_DISTANCE)

    def calculateLinePolygon(self, selectionOffset, line: QLineF):
        nPolygon = QPolygonF()
        radAngle = line.angle() * math.pi / 180.0
        dx = selectionOffset * math.sin(radAngle)
        dy = selectionOffset * math.cos(radAngle)
        offset1 = QPointF(dx, dy)
        offset2 = QPointF(-dx, -dy)
        nPolygon.append(line.p1() + offset1)
        nPolygon.append(line.p1() + offset2)
        nPolygon.append(line.p2() + offset2)
        nPolygon.append(line.p2() + offset1)
        nPolygon.append(line.p1() + offset1)
        return nPolygon

    def updateSelectionPolygon(self):
        self.selectionPath.clear()
        offset = 5

        for i in range(len(self.linePath) - 1):
            p1 = self.linePath[i]
            p2 = self.linePath[i+1]
            self.selectionPath.addPolygon(self.calculateLinePolygon(offset, QLineF(p1, p2)))

    def shape(self):
        return self.selectionPath

    # -----------------------------------
    # Visualization
    def paint(self, painter, option, widget):
        if False == self.isSelected():
            painter.setPen(Qt.SolidLine)
        else:
            painter.setPen(Qt.DashLine)

        for i in range(len(self.linePath) - 1):
            painter.drawLine(self.linePath[i], self.linePath[i+1])

        # NOTE: Debug code
        if self.debugShowSelectionPolygon:
            painter.setPen(QPen("red"))
            painter.drawPath(self.selectionPath)

    # -----------------------------------
    # Connected elements
    def beginConnection(self, fromElement, pos):
        print(f"{fromElement}, {pos}")
        self.removeConnection()
        self.fromElement = fromElement
        self.currentMovePos = pos
        self.recalculateLine()

    def isConnecting(self):
        return (self.fromElement is not None) and (self.toElement is None)

    def moveConnectionTo(self, pos):
        if self.isConnecting():
            self.currentMovePos = pos
            self.recalculateLine()
        else:
            print("WARNING: unexpected moveConnectionTo call")

    def connectElements(self, fromElement, toElement):
        self.removeConnection()
        self.fromElement = fromElement
        self.toElement = toElement
        self.recalculateLine()
        self.fromElement.geometryChanged.connect(self.recalculateLine)
        self.toElement.geometryChanged.connect(self.recalculateLine)

    def removeConnection(self):
        if self.fromElement is not None:
            try:
                self.fromElement.geometryChanged.disconnect(self.recalculateLine)
            except:
                print("from not connected")
        if self.toElement is not None:
            try:
                self.toElement.geometryChanged.disconnect(self.recalculateLine)
            except:
                print("to not connected")
        self.fromElement = None
        self.toElement = None
        # self.recalculateLine()

    def findStartingPointFromPoint(self, rectFrom, pointTo):
        pointFrom = QPointF()

        if (rectFrom.bottom() <= pointTo.y()):
            if rectFrom.right() <= pointTo.x():
                # print("1")
                pointFrom = QPointF(rectFrom.right(), rectFrom.center().y())
            elif rectFrom.left() >= pointTo.x():
                # print("3")
                pointFrom = QPointF(rectFrom.left(), rectFrom.center().y())
            else:
                # print("2, 12, 23")
                pointFrom = QPointF(rectFrom.center().x(), rectFrom.bottom())
        elif (rectFrom.top() >= pointTo.y()):
            if rectFrom.right() <= pointTo.x():
                # print("6")
                pointFrom = QPointF(rectFrom.right(), rectFrom.center().y())
            elif rectFrom.left() >= pointTo.x():
                # print("8")
                pointFrom = QPointF(rectFrom.left(), rectFrom.center().y())
            else:
                # print("7, 67, 78")
                pointFrom = QPointF(rectFrom.center().x(), rectFrom.top())
        elif (rectFrom.right() <= pointTo.x()):
            # print("4, 14, 46")
            pointFrom = QPointF(rectFrom.right(), rectFrom.center().y())
        elif (rectFrom.left() >= pointTo.x()):
            # print("5, 35, 58")
            pointFrom = QPointF(rectFrom.left(), rectFrom.center().y())
        else:
            # print("overlapping")
            pointFrom = rectFrom.center()

        return pointFrom

    def findStartingPointFromRect(self, rectFrom, rectTo):
        pointFrom = None
        if (rectFrom.bottom() <= rectTo.top()):
            if rectFrom.right() <= rectTo.left():
                # print("1")
                pointFrom = QPointF(rectFrom.right(), rectFrom.center().y())
            elif rectFrom.left() >= rectTo.right():
                # print("3")
                pointFrom = QPointF(rectFrom.left(), rectFrom.center().y())
            else:
                # print("2, 12, 23")
                pointFrom = QPointF(rectFrom.center().x(), rectFrom.bottom())
        elif (rectFrom.top() >= rectTo.bottom()):
            if rectFrom.right() <= rectTo.left():
                # print("6")
                pointFrom = QPointF(rectFrom.right(), rectFrom.center().y())
            elif rectFrom.left() >= rectTo.right():
                # print("8")
                pointFrom = QPointF(rectFrom.left(), rectFrom.center().y())
            else:
                # print("7, 67, 78")
                pointFrom = QPointF(rectFrom.center().x(), rectFrom.top())
        elif (rectFrom.right() <= rectTo.left()):
            # print("4, 14, 46")
            pointFrom = QPointF(rectFrom.right(), rectFrom.center().y())
        elif (rectFrom.left() >= rectTo.right()):
            # print("5, 35, 58")
            pointFrom = QPointF(rectFrom.left(), rectFrom.center().y())
            pointTo = QPointF(rectTo.right(), rectTo.center().y())
        else:
            # print("overlapping")
            pointFrom = rectFrom.center()
        return pointFrom

    def recalculateLine(self, element=None):
        # TODO: decide how to snap transition to items
        if (self.fromElement is not None) and (self.toElement is not None):
            # posFrom = self.fromElement.mapToScene(self.fromElement.pos())
            # posTo = self.fromElement.mapToScene(self.toElement.pos())
            rectFrom = self.mapRectFromItem(self.fromElement, self.fromElement.elementRect())
            rectTo = self.mapRectFromItem(self.toElement, self.toElement.elementRect())

            # determine relative position of elements agains eachother
            # if can be one of these 8 positions or a combination of them
            #     1 2 3
            #     4 @ 5
            #     6 7 8
            #
            #    12 23
            #  14 @@@ 35
            #     @@@
            #  46 @@@ 58
            #    67 78
            #
            # 1, 3, 6, 8: angle line
            # others: Z-line
            # rectFrom = QRectF()
            # rectTo = QRectF()

            # print(f"from: {rectFrom}, to: {rectTo}")
            # print(f"from.b={rectFrom.bottom()}, to.t={rectTo.top()}")

            if (self.linePath.size() > 0):
                self.linePath.removeFirst()
                self.linePath.removeLast()

            pointFrom = None
            pointTo = None

            if (len(self.linePath) > 0):
                pointFrom = self.findStartingPointFromPoint(rectFrom, self.linePath.first())
                pointTo = self.findStartingPointFromPoint(rectTo, self.linePath.last())
            else:
                pointFrom = self.findStartingPointFromRect(rectFrom, rectTo)
                pointTo = self.findStartingPointFromRect(rectTo, rectFrom)

            self.linePath.insert(0, pointFrom)
            self.linePath.append(pointTo)
        elif (self.fromElement is not None):
            self.linePath.clear()
            rectFrom = self.mapRectFromItem(self.fromElement, self.fromElement.elementRect())
            pointFrom = self.findStartingPointFromPoint(rectFrom, self.currentMovePos)
            self.linePath.append(pointFrom)
            self.linePath.append(self.currentMovePos)
        else:
            self.linePath.clear()
        self.updateBoundingRect()

    # -----------------------------------
    # Grip notifications
    def onGripMoved(self, grip, pos):
        gripIndex = self.findGripIndex(grip)
        if None != gripIndex:
            self.linePath[gripIndex] = pos
            self.recalculateLine()
            self.updateBoundingRect()
        return True

    def onGripDoubleClick(self, grip):
        gripIndex = self.findGripIndex(grip)
        if None != gripIndex:
            grip.setParentItem(None)
            self.linePath.removeAt(gripIndex)
            del self.lineGrips[gripIndex]
            self.updateBoundingRect()

    # def onGripLostFocus(self, grip):
    #     # TODO
    #     return True

    def findGripIndex(self, grip):
        gripIndex = None
        for i in range(len(self.lineGrips)):
            if self.lineGrips[i] == grip:
                gripIndex = i
                break
        return gripIndex

    def mouseDoubleClickEvent(self, event):
        (sel, p, index) = self.isPointOnTheLine(event.pos())

        if None != index:
            e = ElementGripItem(self)
            e.setPos(p)
            self.linePath.insert(index + 1, p)
            self.lineGrips.insert(index + 1, e)

        return super().mouseDoubleClickEvent(event)

    # -----------------------------------
    # Utils
    def isPointOnTheLine(self, pos):
        intersects = False
        linePoint = QPointF()
        itPrevPoint = None

        cornersCount = len(self.linePath)
        for i in range(cornersCount):
            if (QLineF(self.linePath[i], pos).length() <= self.SELECTION_DISTANCE):
                # Is pressed point close enough of the grip-point
                intersects = True
                break
            else:
                if (i < self.linePath.count() - 1):
                    line = QLineF(self.linePath[i], self.linePath[i + 1])
                    # Is pressed point close enough of the line
                    line2 = QLineF(pos, pos + QPointF(self.SELECTION_DISTANCE, self.SELECTION_DISTANCE))
                    line2.setAngle(line.angle() + 90)
                    (intersection, linePoint) = line.intersects(line2)

                    if QLineF.BoundedIntersection == intersection:
                        intersects = True
                    else:
                        line2.setAngle(line.angle() - 90)
                        (intersection, linePoint) = line.intersects(line2)
                        intersects = (QLineF.BoundedIntersection == intersection)

                    if True == intersects:
                        itPrevPoint = i
                        break

        return (intersects, linePoint, itPrevPoint)

File: idecpp/hsm_ide/elements/HsmTransition.cpp
#include "HsmTransition.hpp"

#include <QGraphicsSceneMouseEvent>
#include <QPainter>
#include <QPainterPath>
#include <QPointF>
#include <QPolygonF>
#include <cmath>
#include <QDebug>
#include "private/ElementGripItem.hpp"

HsmTransition::HsmTransition()
    : HsmElement(HsmElementType::TRANSITION)
    , mFromElement(nullptr)
    , mToElement(nullptr)
    , mDebugShowSelectionPolygon(false) {
    qDebug() << "CREATE: HsmTransition: " << this;
    // NOTE: there are no grips at the beginning and and of the poly-line
    // TODO: fix me
    // self.lineGrips = [None, None]

    setFlag(QGraphicsItem::ItemIsMovable, false);
    setFlag(QGraphicsItem::ItemIsSelectable, true);
    setZValue(0);
    // setAcceptHoverEvents(true);
}

HsmTransition::~HsmTransition() {
    qDebug() << "DELETE: HsmTransition: " << this;
}

void HsmTransition::hoverMoveEvent(QGraphicsSceneHoverEvent *event) {
    qDebug() << "TRANSITION: hoverMoveEvent";
}

void HsmTransition::updateBoundingRect(const QRectF& newRect) {
    prepareGeometryChange();
    updateSelectionPolygon();

    mOuterRect = mLinePath.boundingRect().normalized().adjusted(-static_cast<int>(SELECTION_DISTANCE),
                                                                -static_cast<int>(SELECTION_DISTANCE),
                                                                static_cast<int>(SELECTION_DISTANCE),
                                                                static_cast<int>(SELECTION_DISTANCE));
}

QPolygonF HsmTransition::calculateLinePolygon(int selectionOffset, const QLineF& line) {
    const double radAngle = line.angle() * M_PI / 180.0;
    const double dx = selectionOffset * std::sin(radAngle);
    const double dy = selectionOffset * std::cos(radAngle);
    const QPointF offset1(dx, dy);
    const QPointF offset2(-dx, -dy);
    QPolygonF nPolygon;

    nPolygon.append(line.p1() + offset1);
    nPolygon.append(line.p1() + offset2);
    nPolygon.append(line.p2() + offset2);
    nPolygon.append(line.p2() + offset1);
    nPolygon.append(line.p1() + offset1);

    return nPolygon;
}

void HsmTransition::updateSelectionPolygon() {
    constexpr int offset = 5;

    mSelectionPath.clear();

    for (int i = 0; i < mLinePath.size() - 1; ++i) {
        QPointF p1 = mLinePath[i];
        QPointF p2 = mLinePath[i + 1];

        mSelectionPath.addPolygon(calculateLinePolygon(offset, QLineF(p1, p2)));
    }
}

QPainterPath HsmTransition::shape() const {
    return mSelectionPath;
}

void HsmTransition::paint(QPainter* painter, const QStyleOptionGraphicsItem* option, QWidget* widget) {
    if (false == isSelected()) {
        painter->setPen(Qt::SolidLine);
    } else {
        painter->setPen(Qt::DashLine);
    }

    for (int i = 0; i < (mLinePath.size() - 1); ++i) {
        painter->drawLine(mLinePath[i], mLinePath[i + 1]);
    }

    // TODO: Debug code. Remove later
    if (true == mDebugShowSelectionPolygon) {
        painter->setPen(QPen("red"));
        painter->drawPath(mSelectionPath);
    }
}

void HsmTransition::beginConnection(HsmElement* fromElement, const QPointF& pos) {
    // print(f"{fromElement}, {pos}")
    removeConnection();
    mFromElement = fromElement;
    mCurrentMovePos = pos;
    recalculateLine();
}

bool HsmTransition::isConnecting() const {
    return (nullptr != mFromElement) && (nullptr == mToElement);
}

void HsmTransition::moveConnectionTo(const QPointF& pos) {
    if (true == isConnecting()) {
        mCurrentMovePos = pos;
        recalculateLine();
    } else {
        qDebug() << "WARNING: unexpected moveConnectionTo call";
    }
}

void HsmTransition::connectElements(HsmElement* fromElement, HsmElement* toElement) {
    removeConnection();
    mFromElement = fromElement;
    mToElement = toElement;
    recalculateLine();
    connect(mFromElement, SIGNAL(geometryChanged(HsmResizableElement*)), this, SLOT(recalculateLine()));
    connect(mToElement, SIGNAL(geometryChanged(HsmResizableElement*)), this, SLOT(recalculateLine()));
}

void HsmTransition::removeConnection() {
    if (mFromElement != nullptr) {
        try {
            QObject::disconnect(mFromElement, SIGNAL(geometryChanged(HsmResizableElement*)), this, SLOT(recalculateLine()));
        } catch (...) {
            qDebug() << "from not connected";
        }
    }
    if (mToElement != nullptr) {
        try {
            QObject::disconnect(mToElement, SIGNAL(geometryChanged(HsmResizableElement*)), this, SLOT(recalculateLine()));
        } catch (...) {
            qDebug() << "to not connected";
        }
    }

    mFromElement = nullptr;
    mToElement = nullptr;
    // recalculateLine();
}

QPointF HsmTransition::findStartingPointFromPoint(const QRectF& rectFrom, const QPointF& pointTo) {
    QPointF pointFrom;

    if (rectFrom.bottom() <= pointTo.y()) {
        if (rectFrom.right() <= pointTo.x()) {
            // cout << "1" << endl;
            pointFrom = QPointF(rectFrom.right(), rectFrom.center().y());
        } else if (rectFrom.left() >= pointTo.x()) {
            // cout << "3" << endl;
            pointFrom = QPointF(rectFrom.left(), rectFrom.center().y());
        } else {
            // cout << "2, 12, 23" << endl;
            pointFrom = QPointF(rectFrom.center().x(), rectFrom.bottom());
        }
    } else if (rectFrom.top() >= pointTo.y()) {
        if (rectFrom.right() <= pointTo.x()) {
            // cout << "6" << endl;
            pointFrom = QPointF(rectFrom.right(), rectFrom.center().y());
        } else if (rectFrom.left() >= pointTo.x()) {
            // cout << "8" << endl;
            pointFrom = QPointF(rectFrom.left(), rectFrom.center().y());
        } else {
            // cout << "7, 67, 78" << endl;
            pointFrom = QPointF(rectFrom.center().x(), rectFrom.top());
        }
    } else if (rectFrom.right() <= pointTo.x()) {
        // cout << "4, 14, 46" << endl;
        pointFrom = QPointF(rectFrom.right(), rectFrom.center().y());
    } else if (rectFrom.left() >= pointTo.x()) {
        // cout << "5, 35, 58" << endl;
        pointFrom = QPointF(rectFrom.left(), rectFrom.center().y());
    } else {
        // cout << "overlapping" << endl;
        pointFrom = rectFrom.center();
    }

    return pointFrom;
}

QPointF HsmTransition::findStartingPointFromRect(const QRectF& rectFrom, const QRectF& rectTo) {
    QPointF pointFrom;
    if (rectFrom.bottom() <= rectTo.top()) {
        if (rectFrom.right() <= rectTo.left()) {
            // cout << "1" << endl;
            pointFrom = QPointF(rectFrom.right(), rectFrom.center().y());
        } else if (rectFrom.left() >= rectTo.right()) {
            // cout << "3" << endl;
            pointFrom = QPointF(rectFrom.left(), rectFrom.center().y());
        } else {
            // cout << "2, 12, 23" << endl;
            pointFrom = QPointF(rectFrom.center().x(), rectFrom.bottom());
        }
    } else if (rectFrom.top() >= rectTo.bottom()) {
        if (rectFrom.right() <= rectTo.left()) {
            // cout << "6" << endl;
            pointFrom = QPointF(rectFrom.right(), rectFrom.center().y());
        } else if (rectFrom.left() >= rectTo.right()) {
            // cout << "8" << endl;
            pointFrom = QPointF(rectFrom.left(), rectFrom.center().y());
        } else {
            // cout << "7, 67, 78" << endl;
            pointFrom = QPointF(rectFrom.center().x(), rectFrom.top());
        }
    } else if (rectFrom.right() <= rectTo.left()) {
        // cout << "4, 14, 46" << endl;
        pointFrom = QPointF(rectFrom.right(), rectFrom.center().y());
    } else if (rectFrom.left() >= rectTo.right()) {
        // cout << "5, 35, 58" << endl;
        pointFrom = QPointF(rectFrom.left(), rectFrom.center().y());
        // QPointF pointTo(rectTo.right(), rectTo.center().y());
    } else {
        // cout << "overlapping" << endl;
        pointFrom = rectFrom.center();
    }

    return pointFrom;
}

void HsmTransition::recalculateLine() {
    // TODO: decide how to snap transition to items
    if ((nullptr != mFromElement) && (nullptr != mToElement)) {
        // posFrom = self.fromElement.mapToScene(self.fromElement.pos())
        // posTo = self.fromElement.mapToScene(self.toElement.pos())
        QRectF rectFrom = mapRectFromItem(mFromElement, mFromElement->elementRect());
        QRectF rectTo = mapRectFromItem(mToElement, mToElement->elementRect());

        // determine relative position of elements against each other
        // if can be one of these 8 positions or a combination of them
        //     1 2 3
        //     4 @ 5
        //     6 7 8
        //
        //    12 23
        //  14 @@@ 35
        //     @@@
        //  46 @@@ 58
        //    67 78
        //
        // 1, 3, 6, 8: angle line
        // others: Z-line
        // rectFrom = QRectF()
        // rectTo = QRectF()

        // cout << "from: " << rectFrom << ", to: " << rectTo << endl;
        // cout << "from.b=" << rectFrom.bottom() << ", to.t=" << rectTo.top() << endl;

        if (mLinePath.size() > 0) {
            mLinePath.removeFirst();
            mLinePath.removeLast();
        }

        QPointF pointFrom;
        QPointF pointTo;

        if (false == mLinePath.isEmpty()) {
            pointFrom = findStartingPointFromPoint(rectFrom, mLinePath.first());
            pointTo = findStartingPointFromPoint(rectTo, mLinePath.last());
        } else {
            pointFrom = findStartingPointFromRect(rectFrom, rectTo);
            pointTo = findStartingPointFromRect(rectTo, rectFrom);
        }

        mLinePath.prepend(pointFrom);
        mLinePath.append(pointTo);
        // qDebug() << "mLinePath: " << mLinePath.size() << ": " << mLinePath;
    } else if (nullptr != mFromElement) {
        QRectF rectFrom = mapRectFromItem(mFromElement, mFromElement->elementRect());
        QPointF pointFrom = findStartingPointFromPoint(rectFrom, mCurrentMovePos);

        mLinePath.clear();
        mLinePath.append(pointFrom);
        mLinePath.append(mCurrentMovePos);
    } else {
        mLinePath.clear();
    }

    updateBoundingRect();
}

// =================================================================================================================
// Grip notifications
bool HsmTransition::onGripMoved(const ElementGripItem* grip, const QPointF& pos) {
    const int gripIndex = findGripIndex(grip);
    // qDebug() << "HsmTransition::onGripMoved: index=" << gripIndex << ", pos=" << pos;

    if (gripIndex >= 0) {
        mLinePath[gripIndex] = pos;
        recalculateLine();
        updateBoundingRect();
    }

    return true;
}

void HsmTransition::onGripDoubleClick(ElementGripItem* grip) {
    printf("HsmTransition::onGripDoubleClick\n");
    // const int gripIndex = findGripIndex(grip);
    auto it = std::find(mLineGrips.begin(), mLineGrips.end(), grip);

    if (it != mLineGrips.end()) {
        grip->setParentItem(nullptr);
        // TODO:validate
        mLinePath.removeAt(it - mLineGrips.begin());
        // TODO: check for memory leaks
        // delete mLineGrips.takeAt(gripIndex);
        delete grip;
        mLineGrips.erase(it);
        updateBoundingRect();
    }
}

// # def onGripLostFocus(self, grip):
// #     # TODO
// #     return True

int HsmTransition::findGripIndex(const ElementGripItem* grip) {
    int gripIndex = -1;

    // TODO: use foreach or std::find
    for (int i = 0; i < mLineGrips.size(); ++i) {
        if (mLineGrips[i] == grip) {
            gripIndex = i;
            break;
        }
    }

    return gripIndex;
}

void HsmTransition::mouseDoubleClickEvent(QGraphicsSceneMouseEvent* event) {
    qDebug() << "HsmTransition::mouseDoubleClickEvent";
    std::tuple<bool, QPointF, int> result = isPointOnTheLine(event->pos());

    if (true == std::get<0>(result)) {
        // TODO: use smart pointer
        ElementGripItem* grip = new ElementGripItem(this);
        QPointF gripPos = mapToScene(std::get<1>(result));

        grip->init();
        grip->setPos(std::get<1>(result));
        mLinePath.insert(std::get<2>(result) + 1, std::get<1>(result));
        // TODO: validate
        mLineGrips.insert(mLineGrips.begin() + std::get<2>(result) + 1, grip);
    }

    QGraphicsItem::mouseDoubleClickEvent(event);
}

// =================================================================================================================
// Utils
std::tuple<bool, QPointF, int> HsmTransition::isPointOnTheLine(const QPointF& pos) {
    bool intersects = false;
    QPointF linePoint;
    int prevPointIndex = -1;
    const int cornersCount = mLinePath.size();

    for (int i = 0; i < cornersCount; ++i) {
        if (QLineF(mLinePath[i], pos).length() <= SELECTION_DISTANCE) {
            // Is pressed point close enough of the grip-point
            intersects = true;
            break;
        } else {
            if (i < mLinePath.count() - 1) {
                QLineF line(mLinePath[i], mLinePath[i + 1]);
                // Is pressed point close enough of the line
                QLineF line2(pos, pos + QPointF(SELECTION_DISTANCE, SELECTION_DISTANCE));

                line2.setAngle(line.angle() + 90);

                QLineF::IntersectionType intersection = line.intersects(line2, &linePoint);

                if (QLineF::BoundedIntersection == intersection) {
                    intersects = true;
                } else {
                    line2.setAngle(line.angle() - 90);
                    intersection = line.intersects(line2, &linePoint);
                    intersects = (QLineF::BoundedIntersection == intersection);
                }

                if (true == intersects) {
                    prevPointIndex = i;
                    break;
                }
            }
        }
    }

    return std::make_tuple(intersects, linePoint, prevPointIndex);
}
